src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.cs

```diff
@@ -348,12 +355,19 @@ public IEnumerable<ILSequencePoint> GetSequencePointsForMethod(MethodDesc method
             return moduleData.PdbReader.GetSequencePointsForMethod(MetadataTokens.GetToken(ecmaMethod.Handle));
         }
 
-        public IEnumerable<ILLocalVariable> GetLocalVariableNamesForMethod(MethodDesc method)
+        public IEnumerable<ILLocalVariable> GetLocalVariableNamesForMethod(MethodDesc method, MethodIL methodIL)
         {
-            EcmaMethod ecmaMethod = method.GetTypicalMethodDefinition() as EcmaMethod;
+            MethodDesc uninstantiatedMethod = method.GetTypicalMethodDefinition();
+            MethodIL uninstantiatedMethodIL = methodIL.GetMethodILDefinition();
+
+            EcmaMethod ecmaMethod = uninstantiatedMethod as EcmaMethod;
             if (ecmaMethod == null)
                 return null;
 
+            // If IL is not backed by EcmaMethodIL, the body was generated by the compiler and the PDB won't match.
+            if (!(uninstantiatedMethodIL is EcmaMethodIL))
```

Comment:
I added a new virtual to `MethodIL` that describes the originating `MethodDesc`. I kept the other one I introduced because it's still useful to be able to get to the definition `MethodIL` for a particular `InstantiatedMethodIL`.

(It will be useful once I start the work to do the disassembly - I'm postponing that to a later pull request.).